extensions/amp-a4a/0.1/real-time-config-manager.js

```diff
@@ -0,0 +1,253 @@
+          // is not a required feature.
+          url, {credentials: 'include'}).then(res => {
+            return res.text().then(text => {
+              const rtcTime = Date.now() - rtcStartTime;
+              // An empty text response is allowed, not an error.
+              if (!text) {
+                return {rtcTime, callout};
+              }
+              const response = tryParseJson(text);
+              return response ? {response, rtcTime, callout} :
+              buildErrorResponse_(
+                  RTC_ERROR_ENUM.MALFORMED_JSON_RESPONSE, callout, rtcTime);
+            });
+          })).catch(unusedError => {
+            return buildErrorResponse_(RTC_ERROR_ENUM.NETWORK_FAILURE,
+                callout, Date.now() - rtcStartTime);
+          });
+}
+
+/**
+ * Attempts to parse the publisher-defined RTC config off the amp-ad
+ * element, then validates that the rtcConfig exists, and contains
+ * an entry for either vendor URLs, or publisher-defined URLs. If the
+ * config contains an entry for timeoutMillis, validates that it is a
+ * number, or converts to a number if number-like, otherwise overwrites
+ * with the default.
+ * IMPORTANT: If the rtcConfig is invalid, RTC is aborted, and the ad
+ *   request continues without RTC.
+ * @param {!Element} element
+ * @return {?Object}
+ * @visibleForTesting
+ */
+export function validateRtcConfig_(element) {
+  const defaultTimeoutMillis = 1000;
+  const rtcConfig = tryParseJson(
+      element.getAttribute('rtc-config'));
+  if (!rtcConfig) {
+    return null;
+  }
+  let timeout;
+  try {
+    user().assert(rtcConfig['vendors'] || rtcConfig['urls'],
+        'RTC Config must specify vendors or urls');
+    if (!Object.keys(rtcConfig['vendors'] || {}).length
+       && !(rtcConfig['urls'] || []).length) {
+      return null;
+    }
+    Object.keys(rtcConfig).forEach(key => {
+      switch (key) {
+        case 'vendors':
+          user().assert(isObject(rtcConfig[key]), 'RTC invalid vendors');
+          break;
+        case 'urls':
+          user().assert(isArray(rtcConfig[key]), 'RTC invalid urls');
+          break;
+        case 'timeoutMillis':
+          timeout = parseInt(rtcConfig[key], 10);
+          if (isNaN(timeout)) {
+            user().warn(TAG, 'Invalid RTC timeout is NaN, ' +
+                        `using default timeout ${defaultTimeoutMillis}ms`);
+          } else if (timeout >= defaultTimeoutMillis || timeout < 0) {
+            timeout = null;
+            user().warn(TAG, `Invalid RTC timeout: ${timeout}ms, ` +
+                        `using default timeout ${defaultTimeoutMillis}ms`);
+          }
+          break;
+        default:
+          user().warn(TAG, `Unknown RTC Config key: ${key}`);
+          break;
+      }
+    });
+  } catch (unusedErr) {
+    // This error would be due to the asserts above.
+    return null;
+  }
+
+  rtcConfig['timeoutMillis'] = timeout || defaultTimeoutMillis;
```

Comment:
Do we want to allow a timeout of 0?  If so, this is not causing it to be used given timeout is falsy.  Please add a test for this.

Reply:
done